### ISSUE TITLE
FAIL: datalad.distributed.tests.test_ria_basics.test_remote_layout('datalad-test',)

### ISSUE THREAD

**Issue by mih** [datalad/datalad#5302]
```
Traceback (most recent call last):
2017  File "/Users/appveyor/venv3.8.6/lib/python3.8/site-packages/nose/case.py", line 198, in runTest
2018    self.test(*self.arg)
2019  File "/Users/appveyor/venv3.8.6/lib/python3.8/site-packages/datalad/tests/utils.py", line 291, in _wrap_skip_ssh
2020    return func(*args, **kwargs)
2021  File "/Users/appveyor/venv3.8.6/lib/python3.8/site-packages/datalad/tests/utils.py", line 263, in _wrap_skip_if
2022    return func(*args, **kwargs)
2023  File "/Users/appveyor/venv3.8.6/lib/python3.8/site-packages/datalad/tests/utils.py", line 772, in _wrap_with_tempfile
2024    return t(*(arg + (filename,)), **kw)
2025  File "/Users/appveyor/venv3.8.6/lib/python3.8/site-packages/datalad/tests/utils.py", line 772, in _wrap_with_tempfile
2026    return t(*(arg + (filename,)), **kw)
2027  File "/Users/appveyor/venv3.8.6/lib/python3.8/site-packages/datalad/tests/utils.py", line 772, in _wrap_with_tempfile
2028    return t(*(arg + (filename,)), **kw)
2029  File "/Users/appveyor/venv3.8.6/lib/python3.8/site-packages/datalad/distributed/tests/test_ria_basics.py", line 289, in _test_remote_layout
2030    assert_equal(len(ds.repo.whereis('one.txt')), len(whereis) + 1)
2031AssertionError: 2 != 3
```

**Comment by mih** [comment]
I cannot replicate the failure on the macos appveyor workers (anymore). Closing for now, maybe it resurfaces.

**Comment by mih** [comment]
And here it is: https://ci.appveyor.com/project/mih/datalad/builds/37191972/job/7ollki4j3m3l6ocj#L2018

Sadly, "random" behavior...

**Comment by mih** [comment]
This test made negative news before: https://github.com/datalad/datalad/issues/5145

**Comment by mih** [comment]
When I go into a failing run. I see the following `fsck` output for the final and unsuccessful `fsck`:

```
(venv3.8.6) worker-628-001:datalad_temp__test_remote_layouth5lsnix8 appveyor$ git annex fsck -f archive
fsck one.txt 
  Remote doesn't report any dataset tree version. Consider upgrading datalad or add a fitting 'ria-layout-version' file at the RIA store's root.Treating remote as read-only in order toprevent damage by putting things into an unknown version of the target layout. You can overrule this by setting 'annex.ora-remote.<name>.force-write=true'.

  Remote doesn't report any object tree version. Consider upgrading datalad or add a fitting 'ria-layout-version' file at the remote dataset root. Treating remote as read-only in order toprevent damage by putting things into an unknown version of the target layout. You can overrule this by setting 'annex.ora-remote.<name>.force-write=true'.
ok
fsck subdir/two ok
```

**Comment by mih** [comment]
Ok, I take the spurios back. I do not know how that test could have passed before. At minimum the 'archive' store needs to be initialized properly to prevent the special remote from treating it as half-broken:

```
diff --git a/datalad/distributed/tests/test_ria_basics.py b/datalad/distributed/tests/test_ria_basics.py
index 835ce1319..db939bfe5 100644
--- a/datalad/distributed/tests/test_ria_basics.py
+++ b/datalad/distributed/tests/test_ria_basics.py
@@ -273,7 +273,11 @@ def _test_remote_layout(host, dspath, store, archiv_store):
 
         # we can simply pack up the content of the remote into a
         # 7z archive and place it in the right location to get a functional
-        # archive remote
+        # but we need to put a placeholder in first that identifies layout
+        # version and object tree version to get a functional remote
+        create_store(io, archiv_store, '1')
+        create_ds_in_store(io, archiv_store, ds.id, '2', '1')
+
         whereis = ds.repo.whereis('one.txt')
         dsgit_dir, archive_dir, dsobj_dir = \
             get_layout_locations(1, archiv_store, ds.id)
```

However, even after such a change, the special remote does indeed report the keys in the archive.7z as unavailable, although SSH login works fine and `7z` is available in the remote shell.

Given that the sister test without SSH is working, the culprit seems to be either in the special remote implementation, or some specific interaction with the SSH setup on the worker.

**Comment by mih** [comment]
I am adding this to the list of must-solve issues for 0.14. If this is not a pure test setup issue, this kind of failure is a serious problem.

**Comment by bpoldrack** [comment]
> I do not know how that test could have passed before. At minimum the 'archive' store needs to be initialized properly to prevent the special remote from treating it as half-broken

Nope. Store should be there already at that point.

Looking into it. Suspect something is wrong with the path to the store (or its URL).

**Comment by bpoldrack** [comment]
Update: So, the mystery why it passed at all is not a mystery. Read access to correctly structured store that just fails to declare its layout is fine.

The actual issue is, that within ORA via SSH, the output from the command ran to check what's in the archive appears to be empyt (although manually checking that archive on that machine via 7z works just fine). So, something doesn't work regarding the execution of `7z l` via SSH.

I'll keep digging.

**Comment by bpoldrack** [comment]
Aaand it's `command not found` from within the SSH shell that's actually used by ORA. Not clear to me yet, since simple `ssh datalad-test` on that VM ends up in an env with `7z` in PATH.

**Comment by bpoldrack** [comment]
Another piece of the puzzle:

`echo $PATH` from within ORA's shell vs from within `ssh datalad-test` on that VM (as user `appveyor`):

`/usr/bin:/bin:/usr/sbin:/sbin`

vs

`/Library/Frameworks/Mono.framework/Home/bin:/Users/appveyor/.nvm/versions/node/v12.20.0/bin:/Users/appveyor/.gvm/pkgsets/go1.15.6/global/bin:/Users/appveyor/.gvm/gos/go1.15.6/bin:/Users/appveyor/.gvm/pkgsets/go1.15.6/global/overlay/bin:/Users/appveyor/.gvm/bin:/Users/appveyor/.gvm/bin:/Users/appveyor/.fastlane/bin:/Users/appveyor/.rvm/gems/ruby-2.6.5/bin:/Users/appveyor/.rvm/gems/ruby-2.6.5@global/bin:/Users/appveyor/.rvm/rubies/ruby-2.6.5/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/Users/appveyor/.rvm/bin:/Users/appveyor/vcpkg:/usr/local/share/dotnet`


Note, that in the former `/usr/local/bin` is missing (and that's where `7z` actually is).
However, ORA's shell still lets `whoami` respond `appveyor`.

**Comment by mih** [comment]
That seems to be the default setup for `sshd` on macos. I think the key issue is that a user has no idea that any of that is happening. Why is the ORA remote explicitly disabling the check whether a command has failed on the remote? It seems that it has the capability to detect and to report. It just turns it off.

```py
def in_archive(...):
       ...
        # Note: Currently relies on file_path not showing up in case of failure
        # including non-existent archive. If need be could be more sophisticated
        # and called with check=True + catch RemoteCommandFailedError
        out = self._run(cmd, no_output=False, check=False)
```

**Comment by bpoldrack** [comment]
> That seems to be the default setup for sshd on macos

Yeah, something like this. But:

Not even this does reproduce it:

```
ssh datalad-test "echo $PATH"

/Users/appveyor/venv3.8.6/bin:/Library/Frameworks/Mono.framework/Home/bin:/Users/appveyor/.nvm/versions/node/v12.20.0/bin:/Users/appveyor/.gvm/pkgsets/go1.15.6/global/bin:/Users/appveyor/.gvm/gos/go1.15.6/bin:/Users/appveyor/.gvm/pkgsets/go1.15.6/global/overlay/bin:/Users/appveyor/.gvm/bin:/Users/appveyor/.gvm/bin:/Users/appveyor/.fastlane/bin:/Users/appveyor/.rvm/gems/ruby-2.6.5/bin:/Users/appveyor/.rvm/gems/ruby-2.6.5@global/bin:/Users/appveyor/.rvm/rubies/ruby-2.6.5/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/Users/appveyor/.rvm/bin:/Users/appveyor/vcpkg:/usr/local/share/dotnet
```

> Why is the ORA remote explicitly disabling the check whether a command has failed on the remote?

By design not as easy as it looks and I don't have a general solution (we execute more than just 7z that way) yet.

**Comment by bpoldrack** [comment]
I have added a probe for availability of `7z`, though. https://github.com/datalad/datalad/pull/5303/commits/9486807bf905bf2e19e1bb281b6f5e973b744d59

```
appveyor$ git annex info archive
uuid: 3ada414d-43f5-42a1-8b77-cb31fbe2c58c
description: [archive]
trust: semitrusted
remote: archive
cost: 200.0
type: external
externaltype: ora
7z: not available
storage_host: datalad-test
store_base_path: /Users/appveyor/DLTMP/datalad_temp__test_remote_layoutjl8s0cf_
encryption: none
chunking: none
remote annex keys: 0
remote annex size: 0 bytes
```

**Comment by mih** [comment]
> I have added a probe for availability of `7z`, though. [9486807](https://github.com/datalad/datalad/commit/9486807bf905bf2e19e1bb281b6f5e973b744d59)

The patch does not say why it is done this way, but running `which 7z` in the context of local execution already rules out any execution on windows. Why not just `7z` -- which gives `exit 0`. Or is `which` is necessary, use the windows `where` command.

I know that execution on windows is broken ATM, but as far as I know that is only due to some unaddressed minor issues, and not by design. It makes sense to not complicate the situation further. 

The fact that these tests are not running on windows makes such issues invisible without constant attention.

**Comment by bpoldrack** [comment]
Reason: Had the intention to use the output to store the path and use absolute calls, so we could replace it by any mechanism to locate it and not rely on PATH, since one actually shouldn't in such a non-interactive session. However, for now the exit code would suffice, I agree.

**Comment by bpoldrack** [comment]
Note re `which`/`where`: `type` or `command` may be cross-platform solutions.

**Comment by mih** [comment]
Fixed by https://github.com/datalad/datalad/pull/5303/commits/9104ffb6460bb013b2da34856f16bee69922e3cb from #5303

**Comment by mih** [comment]
Addressed with #5330 

https://ci.appveyor.com/project/mih/datalad/builds/37278054/job/3yv6exm9q91pe0v6#L759
